From ee85b3b59c9a0fc9714f0edffa7f45b05a4e60c0 Mon Sep 17 00:00:00 2001
From: Gregor Herburger <gregor.herburger@tq-group.com>
Date: Tue, 22 Mar 2022 14:10:39 +0100
Subject: [PATCH] mfd: tqmx86: Add platform_data for smarc with 14 gpios

In PLD revision >= 2 all 14 gpios are useable as interrupts for smarc
modules. So check PLD revision and use corresponding platform data.

Signed-off-by: Gregor Herburger <gregor.herburger@tq-group.com>
---
 drivers/mfd/tqmx86.c | 19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

diff --git a/drivers/mfd/tqmx86.c b/drivers/mfd/tqmx86.c
index ea0c0f37346c3..c54f3904c696c 100644
--- a/drivers/mfd/tqmx86.c
+++ b/drivers/mfd/tqmx86.c
@@ -115,6 +115,11 @@ static const struct tqmx86_gpio_platform_data gpio_platform_data_smarc21 = {
 	.dir_input_mask = 0x3fc0, /* 6 outputs, 8 inputs */
 };
 
+static const struct tqmx86_gpio_platform_data gpio_platform_data_smarc21_full = {
+	.ngpio = 14,
+	.dir_input_mask = 0x3fc0, /* 6 outputs, 8 inputs */
+};
+
 enum tqmx86_cells {
 	TQMX86_WDT,
 	TQMX86_GPIO,
@@ -136,7 +141,7 @@ static struct mfd_cell tqmx86_devs[] = {
 };
 
 static const struct tqmx86_gpio_platform_data *
-tqmx86_board_id_to_gpio_pdata(u8 board_id, u8 sauc)
+tqmx86_board_id_to_gpio_pdata(u8 board_id, u8 sauc, u8 pld_rev)
 {
 	switch (board_id) {
 	case TQMX86_REG_BOARD_ID_E38M:
@@ -157,7 +162,9 @@ tqmx86_board_id_to_gpio_pdata(u8 board_id, u8 sauc)
 			? &gpio_platform_data_come
 			: &gpio_platform_data_smarc20;
 	case TQMX86_REG_BOARD_ID_E40S:
-		return &gpio_platform_data_smarc21;
+		return (pld_rev > 1)
+			? &gpio_platform_data_smarc21_full
+			: &gpio_platform_data_smarc21;
 	default:
 		return ERR_PTR(-EINVAL);
 	}
@@ -229,7 +236,7 @@ static int tqmx86_board_id_to_clk_rate(struct device *dev, u8 board_id)
 static int tqmx86_probe(struct platform_device *pdev)
 {
 	const struct tqmx86_gpio_platform_data *gpio_pdata;
-	u8 board_id, sauc, rev, i2c_det, io_ext_int_val;
+	u8 board_id, sauc, rev, i2c_det, io_ext_int_val, pcb_rev, pld_rev;
 	struct device *dev = &pdev->dev;
 	u8 gpio_irq_cfg, readback;
 	const char *board_name;
@@ -262,10 +269,12 @@ static int tqmx86_probe(struct platform_device *pdev)
 	sauc = ioread8(io_base + TQMX86_REG_SAUC);
 	board_name = tqmx86_board_id_to_name(board_id, sauc);
 	rev = ioread8(io_base + TQMX86_REG_BOARD_REV);
+	pcb_rev = rev >> 4;
+	pld_rev = rev & 0xf;
 
 	dev_info(dev,
 		 "Found %s - Board ID %d, PCB Revision %d, PLD Revision %d\n",
-		 board_name, board_id, rev >> 4, rev & 0xf);
+		 board_name, board_id, pcb_rev, pld_rev);
 
 	/*
 	 * The I2C_DETECT register is in the range assigned to the I2C driver
@@ -301,7 +310,7 @@ static int tqmx86_probe(struct platform_device *pdev)
 			return err;
 	}
 
-	gpio_pdata = tqmx86_board_id_to_gpio_pdata(board_id, sauc);
+	gpio_pdata = tqmx86_board_id_to_gpio_pdata(board_id, sauc, pld_rev);
 	if (IS_ERR(gpio_pdata))
 		return PTR_ERR(gpio_pdata);
 
